designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+
+#### Single Reason Approach: One Reason per Budget
+
+### Proposed Spec
+In this approach, each budget entry specifies a single reason for disruption.
+
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // +kubebuilder:validation:Enum:={"all", "default","consolidation","expired","empty","drift"}
+      // +optional
+      Reason string `json:"reason,omitempty" hash:"ignore"`
+      // Nodes dictates the maximum number of NodeClaims owned by this NodePool
+      // that can be terminating at once. This is calculated by counting nodes that
+      // have a deletion timestamp set, or are actively being deleted by Karpenter.
+      // This field is required when specifying a budget.
+      // This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
+      // checking for int nodes for IntOrString nodes.
+      // Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
+      // +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
+      // +kubebuilder:default:="10%"
+      Nodes string `json:"nodes" hash:"ignore"`
+      // Schedule specifies when a budget begins being active, following
+      // the upstream cronjob syntax. If omitted, the budget is always active.
+      // Timezones are not supported.
+      // This field is required if Duration is set.
+      // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$`
+      // +optional
+      Schedule *string `json:"schedule,omitempty" hash:"ignore"`
+      // Duration determines how long a Budget is active since each Schedule hit.
+      // Only minutes and hours are accepted, as cron does not work in seconds.
+      // If omitted, the budget is always active.
+      // This is required if Schedule is set.
+      // This regex has an optional 0s at the end since the duration.String() always adds
+      // a 0s at the end.
+      // +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
+      // +kubebuilder:validation:Type="string"
+      // +optional
+      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+
+```
+##### Example
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      reason: drifted 
+      nodes: 10
+    # For all other reasons , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+
+```
+
+
+#### Pros and Cons
+👍 Simplicity and Clarity: Offers a straightforward and clear mapping of budget to disruption reason.
+👎 Increased Configuration Overhead: Requires duplicating settings for multiple reasons, increasing setup complexity.
```

Comment:
Making it easier to reach the 50 limit as well.

Reply:
To be fair, we could always bump the limit. Having multiple reasons allows users to read their budgets a little easier. If we expect users to lump expiration and drift together, I feel like it makes sense to make this a list.